### PR TITLE
SVCPLAN-5268: BREAKING - update to support stateful nodes

### PR DESCRIPTION
**Pull request by jakerundall**
Change configure_lustre_command (LNet configure command):
- begin by unmounting Lustre and unloading Lustre kernel modules (because these may be activated as part of bootup on stateful nodes)
- add the last system boot time (who -b) into the lnet_trigger_file

Change the condition which causes Exec of configure_lustre_command (causes Lustre/LNet to be (re)configured):
- (re)configure if lnet_trigger_file is missing OR if the system boot time in lnet_trigger_file is different than the actual system boot time (i.e., if the system has been rebooted)

Add warning to README.md about how to handle this breaking change.

### CODE REVIEW

**Comment by jakerundall** [comment]
Developed and tested on some mg nodes at first (which are stateless) but also fully tested on service-backup04, which is stateful.

I'll also add that I verified consistent output of `uptime -s` on the various `control-test-*` nodes, which cover EL 7/8/9, using local, ssh, and xdsh contexts.

File: README.md [hunk 1]
```diff
@@ -36,6 +36,30 @@ include ::profile_lustre::lnet_router
...
+system boot time in the file. On subsequent Puppet runs, if the file is missing OR if the system
+boot time recorded in the file is different than the actual system boot time, then Puppet will
+necessarily attempt to unmount Lustre and unload Lustre/LNet kernel modules, and then (re)configure
+LNet.
+
+(This change in the trigger file is necessary to support stateful systems.)
+
+When introducing this updated module into a cluster with booted production systems admins may want to
+"preseed" each node's system boot time into their trigger file to prevent Puppet from (unmounting Lustre
+and unloading Lustre/LNet kernel modules and) reconfiguring LNet. This can be done by running these commands
```
Comment (billglick) [inline README.md:50]: Nice. I hadn't considered this change possibly causing issues for stateless nodes.

File: README.md [hunk 2]
```diff
@@ -36,6 +36,30 @@ include ::profile_lustre::lnet_router
...
+# update the message in the trigger file, to avoid potential confusion due to the behavior change
+`echo "If this file is deleted or contains an older system boot time Puppet will unmount Lustre and reconfigure LNet" > /etc/lnet.trigger`
+# put the system's boot time in the trigger file
+`uptime -s >> /etc/lnet.trigger'
+
+NOTE: Time-related commands can sometimes produce varying output formats depending on the shell context.
+E.g., `who -b` produces different output when run via xdsh as compared to plain ssh or locally (and this
+is true even when Puppet is run via xdsh). But `uptime -s` produces consistent output in all of these
+contexts. That being said, if admins desire to use some other parallel shell command they should be sure
+aware that varying output format could produce issues and should test carefully.
```
Comment (billglick) [inline README.md]: I had trouble parsing this sentence. I think there is an extra `sure` in the middle of it.

```suggestion
contexts. That being said, if admins desire to use some other parallel shell command they should be
aware that varying output format could produce issues and should test carefully.
```

File: manifests/module.pp [hunk 3]
```diff
@@ -138,13 +138,17 @@
...
+    modprobe lnet && lnetctl lnet configure && lnetctl import ${lnet_conf_file} && \
+    echo 'If this file is deleted or contains an older system boot time Puppet will unmount Lustre and reconfigure LNet' > ${lnet_trigger_file} && \
+    uptime -s >> ${lnet_trigger_file}"
+
+  ## configure LNet, but only if the trigger file is missing or has a different last reboot time
   exec { 'configure_lustre':
     command => $configure_lustre_command,
-    creates => $lnet_trigger_file,
-    unless  => "test -f ${lnet_trigger_file}",
+    unless  => "test -f ${lnet_trigger_file} && egrep -q \"$(uptime -s)\" ${lnet_trigger_file}",
```
Comment (billglick) [inline manifests/module.pp]: I don't think you still need to test if the `lnet_trigger_file` exists. I'm pretty sure the egrep command will fail if the file doesn't exist. It doesn't hurt to leave both configured, but it does make the unless command a bit harder to read.

**Review by billglick — APPROVED** [review]
Looks good. I did add a few comments and suggestions, but feel free to proceed.

File: README.md (hunk 2 above)
Comment (jakerundall) [reply README.md]: Done, I also changed the end of that from "and should test carefully" to "and test carefully".

File: manifests/module.pp (hunk 3 above)
Comment (jakerundall) [reply manifests/module.pp]: You're right. I've updated to remove the unnecessary check for file and retested the trigger conditions on a Magnus node.

**Comment by jakerundall** [comment]
I think we're ready for another look at your convenience. I'll rebase when it comes time to merge.

**Review by billglick — APPROVED** [review]
Looks good

**Pull request merged**